src/compiler/checker.ts

```diff
@@ -11904,6 +11909,21 @@ namespace ts {
             return true;
         }
 
+        function isAritySmaller(sourceType: Type, target: Expression) {
+            if (isFunctionExpressionOrArrowFunction(target) && isFunctionType(sourceType)) {
+                let targetParameterCount = 0;
+                for (; targetParameterCount < target.parameters.length; targetParameterCount++) {
+                    const param = target.parameters[targetParameterCount];
+                    if (param.initializer || param.questionToken || param.dotDotDotToken || isJSDocOptionalParameter(param)) {

```

Comment:
This skips some additional overloads. It is not bad but it is worse than what we do today. Here's the test failure:

```ts
    function foo6<T>(cb: { (x: T): string; (x: T, y?: T): string }) {
        return cb;
    }
    var r11 = foo6(<T>(x: T, y?: T) => ''); // any => string (+1 overload)
```